src/compiler/lowering/MatMulOpLowering.cpp

```diff
@@ -14,198 +14,564 @@
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <cstdint>
 #include <memory>
+#include <stdexcept>
 #include <utility>
 #include <vector>
 
-#include "compiler/utils/CompilerUtils.h"
 #include "compiler/utils/LoweringUtils.h"
+#include "hwloc.h"
 #include "ir/daphneir/Daphne.h"
 #include "ir/daphneir/Passes.h"
-#include "llvm/ADT/APFloat.h"
-#include "llvm/ADT/ArrayRef.h"
 #include "mlir/Conversion/AffineToStandard/AffineToStandard.h"
 #include "mlir/Conversion/ArithToLLVM/ArithToLLVM.h"
 #include "mlir/Conversion/ControlFlowToLLVM/ControlFlowToLLVM.h"
 #include "mlir/Conversion/FuncToLLVM/ConvertFuncToLLVM.h"
-#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
 #include "mlir/Conversion/LLVMCommon/LoweringOptions.h"
 #include "mlir/Conversion/LLVMCommon/TypeConverter.h"
 #include "mlir/Conversion/LinalgToStandard/LinalgToStandard.h"
 #include "mlir/Conversion/MemRefToLLVM/MemRefToLLVM.h"
 #include "mlir/Conversion/SCFToControlFlow/SCFToControlFlow.h"
 #include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/LoopUtils.h"
+#include "mlir/Dialect/Affine/Passes.h"
+#include "mlir/Dialect/Affine/Utils.h"
 #include "mlir/Dialect/Arith/IR/Arith.h"
 #include "mlir/Dialect/Func/Transforms/FuncConversions.h"
 #include "mlir/Dialect/LLVMIR/LLVMDialect.h"
 #include "mlir/Dialect/Linalg/IR/Linalg.h"
 #include "mlir/Dialect/MemRef/IR/MemRef.h"
 #include "mlir/Dialect/SCF/IR/SCF.h"
-#include "mlir/IR/BuiltinDialect.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/BuiltinOps.h"
 #include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/Operation.h"
 #include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Types.h"
 #include "mlir/IR/UseDefLists.h"
 #include "mlir/Pass/Pass.h"
 #include "mlir/Support/LLVM.h"
 #include "mlir/Support/LogicalResult.h"
 #include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+
+namespace mlir {
+#define GEN_PASS_DECL_MATMULOPLOWERINGPASS
+#define GEN_PASS_DEF_MATMULOPLOWERINGPASS
+#include "ir/daphneir/Passes.h.inc"
+} // namespace mlir
 
 using namespace mlir;
 
 static constexpr int ROW = 0;
 static constexpr int COL = 1;
 
-void affineMatMul(mlir::Value &lhs, mlir::Value &rhs, mlir::Value &output,
-                  ConversionPatternRewriter &rewriter, mlir::Location loc,
-                  ArrayRef<int64_t> lhsShape, ArrayRef<int64_t> rhsShape,
-                  mlir::MLIRContext *ctx) {
-    SmallVector<Value, 4> loopIvs;
-
-    // row loop
-    auto rowLoop = rewriter.create<AffineForOp>(loc, 0, lhsShape[ROW], 1);
-    for (Operation &nested : *rowLoop.getBody()) {
-        rewriter.eraseOp(&nested);
+struct LowerMatMulOpOptions {
+  LowerMatMulOpOptions() {}
+  int vec_size_bits{0};
+  bool vectorize{false};
+  bool tile{false};
+  bool useFixedTileSizes{false};
+  int register_size{4 * 4 * 64};
```

Comment:
Which register size is this referring to?

Reply:
It took me a while to really understand what Uday Bondhugula was referring to with this in the blog post. It seems to mean the SIMD vector size x how many vectors there are. I changed the variable name to something that should make this more clear.